src/blockchain/Blockchain.ts

```diff
@@ -80,46 +88,53 @@ export default class Blockchain implements IBlockchainApi {
     return this.api.rpc.chain.subscribeNewHeads(listener)
   }
 
-  public async submitTx(
+  public async signTx(
     identity: Identity,
     tx: SubmittableExtrinsic
-  ): Promise<SubmittableResult> {
-    const accountAddress = identity.address
-    const nonce = await this.getNonce(accountAddress)
+  ): Promise<SubmittableExtrinsic> {
+    const nonce = await this.getNonce(identity.address)
     const signed: SubmittableExtrinsic = identity.signSubmittableExtrinsic(
       tx,
       nonce.toHex()
     )
-    log.info(`Submitting ${tx.method}`)
+    return signed
+  }
 
+  public async submitTx(
+    tx: SubmittableExtrinsic,
+    resolveOn: (result: SubmittableResult) => boolean = (result) =>
+      result.isFinalized
+  ): Promise<SubmittableResult> {
+    log.info(`Submitting ${tx.method}`)
+    let unsubscribe: () => void
     return new Promise<SubmittableResult>((resolve, reject) => {
-      signed
-        .send((result) => {
-          log.info(`Got tx status '${result.status.type}'`)
-
-          if (ErrorHandler.extrinsicFailed(result)) {
-            log.warn(`Extrinsic execution failed`)
-            log.debug(`Transaction detail: ${JSON.stringify(result, null, 2)}`)
-            const extrinsicError: ExtrinsicError =
-              this.errorHandler.getExtrinsicError(result) || ERROR_UNKNOWN
-
-            log.warn(`Extrinsic error ocurred: ${extrinsicError}`)
-            reject(extrinsicError)
-          }
-          if (result.isFinalized) {
-            resolve(result)
-          } else if (result.isError) {
-            reject(
-              new Error(
-                `Transaction failed with status '${result.status.type}'`
-              )
-            )
-          }
+      tx.send((result) => {
+        log.info(`Got tx status '${result.status.type}'`)
+        if (ErrorHandler.extrinsicFailed(result)) {
+          log.warn(`Extrinsic execution failed`)
+          log.debug(`Transaction detail: ${JSON.stringify(result, null, 2)}`)
+          const extrinsicError: ExtrinsicError =
+            this.errorHandler.getExtrinsicError(result) || ERROR_UNKNOWN
```

Comment:
I would love to turn this into a static function, but the error handler is in the way. Any suggestions whether and how we could get rid of this?

Reply:
see 0a4a8127a18d75fa29ab6efdf680569af34d75fa 👏